src/cmap/connection.ts

```diff
@@ -537,6 +538,11 @@ export class Connection extends TypedEventEmitter<ConnectionEvents> {
         }
 
         if (document.ok === 0) {
+          if (options.timeoutContext?.csotEnabled() && document.isMaxTimeExpiredError) {
+            throw new MongoOperationTimeoutError('Server reported a timeout error', {
+              cause: new MongoServerError((object ??= document.toObject(bsonOptions)))
```

Comment:
```suggestion
              cause: new MongoServerError(document.toObject(bsonOptions))
```
Before this, object is only assigned to `null`, what's the reasoning behind adding `object ??= ..` ?

Reply:
This caches the work done by toObject so it is not repeated. Once we've decoded the BSON once we won't again, for command monitoring.

Reply:
In the case of this function, we reassign 'object = null' at the start of every loop, and it's only used once if the error is thrown (since the `throw` statement is not wrapped in a `try-catch`). In my understanding, it doesn't make sense to cache in this way if we're just going to throw the error anyways.

Reply:
You're correct, but it is not something I changed in this PR I would prefer to leave it put because of the consistency it visually provides when looking at the function. This is a holdover until we get the entire driver migrated to using MongoDBResponse, until then this code block is carefully written to prevent us from double serializing every message. 

So that's my feeling, If you feel strongly I will change it

Reply:
I see what you're saying about visual similarity / wanting this error to look the same as the other one. I'm good with keeping it as is.